runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3425,664 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      TR::Node *grandChild = node->getSecondChild()->getFirstChild();
+      TR_ASSERT(grandChild->getSymbolReference() && grandChild->getSymbolReference()->getSymbol()->isClassObject(), "child of second child of direct wrtbar should be class object symbol n%dn (%p)", node->getGlobalIndex(), node);
+      return node->getSecondChild()->getFirstChild(); // the direct second child is JavaLangClass which is loaded from the j9class
+      }
+   else 
+      return node->getFirstChild();
+   }
+
+/*
+   * Fill in the J9JITWatchedStaticFieldData.fieldAddress, J9JITWatchedStaticFieldData.fieldClass for static field
+   * and J9JITWatchedInstanceFieldData.offset for instance field
+   *
+   */
+
+void generateFillInDataBlockSequenceForUnresolvedField (TR::Node *node, J9Method* owningMethod, TR::S390WritableDataSnippet *dataSnippet, bool isWrite, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   TR::SymbolReference *symRef = node->getSymbolReference();
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isStatic = symRef->getSymbol()->getKind() == TR::Symbol::IsStatic;
+
+   TR_RuntimeHelper helperIndex = isWrite? (isStatic ?  TR_jitResolveStaticFieldSetterDirect: TR_jitResolveFieldSetterDirect):
+                                          (isStatic ?  TR_jitResolveStaticFieldDirect: TR_jitResolveFieldDirect);
+
+   TR::S390HelperLinkage *helperLink = static_cast<TR::S390HelperLinkage*>(cg->getLinkage(runtimeHelperLinkage(helperIndex)));
+
+   intptr_t offsetInDataBlock = isStatic ? offsetof(J9JITWatchedStaticFieldData, fieldAddress): offsetof(J9JITWatchedInstanceFieldData, offset);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *unresolvedLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   // need 2 registers for arguments, 2 for entry point and return address register, and 1 for datablock
+   uint8_t numOfConditions = 5;
+
+   if (isStatic) // needs fieldClassReg
+      numOfConditions++;
+
+   TR::RegisterDependencyConditions  *deps =  generateRegisterDependencyConditions(0, numOfConditions, cg);
+   TR::Register *resultReg = NULL;
+   TR::Register *tempReg = cg->allocateRegister();
+   TR::Register *dataBlockReg = cg->allocateRegister();
+
+   deps->addPostCondition(dataBlockReg, TR::RealRegister::AssignAny);
+
+   generateRILInstruction(cg, TR::InstOpCode::LARL, node, dataBlockReg, dataSnippet);
+   generateRILInstruction(cg, TR::InstOpCode::LGFI, node, tempReg, static_cast<int32_t>(-1));
+
+   generateRXInstruction(cg, is64Bit ? TR::InstOpCode::CG : TR::InstOpCode::C, node, tempReg, generateS390MemoryReference(dataBlockReg, offsetInDataBlock, cg));
+
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BNE, node, endLabel); // branch to end if not set
```

Comment:
Instead of the `LGFI`, `CG`, and `BRC` we should be using `generateS390CompareAndBranchInstruction` API which can compare against an immediate value. It will generate the best possible sequence (`CGIJ` for example) instead of this 3 instruction sequence. You then don't need `tempReg` either.

Reply:
We will need to load the data in memory specified by this reference : `generateS390MemoryReference(dataBlockReg, offsetInDataBlock, cg)` in order to use the `generateS390CompareAndBranchInstruction` API.

The relevant changes are here:
https://github.com/eclipse/openj9/pull/4743/commits/f47412decdca59fa6814915ad28e57d5ac2cb837 + https://github.com/eclipse/openj9/pull/4743/commits/1d9fce5a23abacb0a12dbff7a6f919251f4d0fbe